From d4c29e871d6895ba26b148d5c241865723322332 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 7 Apr 2025 18:00:04 +0800
Subject: [PATCH 091/115] mtk: mac80211: refactor BSS color change & collision
 detection

1. add the check for color change to unavailable color
2. fix AP MLD color collision detection on incorrect link.
   When AP MLD receives beacons from other BSSes, the link ID in RX
   status might not be correct because driver is unable to determine
   the interface to handle the beacon and the corresponding link id.
   This change makes sure that the beacon is handled by correct link
   on the band specified in the RX status, and skip the beacon color
   collision detection on the interface if the interface does not
   have link on the band.
3. Reset used_color_bitmpa after color change so that the colors that
   are not in used can be release.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>

We apply the same determination about link_id and band elsewhere to
ensure it functions for the STA interface.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 net/mac80211/cfg.c         |  8 ++++++++
 net/mac80211/ieee80211_i.h |  1 +
 net/mac80211/rx.c          | 25 +++++++++++++++++++------
 3 files changed, 28 insertions(+), 6 deletions(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 3b6f7242..0534bdfd 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -5141,6 +5141,9 @@ static int ieee80211_color_change_finalize(struct ieee80211_link_data *link)
 
 	link->conf->color_change_active = false;
 
+	/* Clear tha map after color change so that colors can be released */
+	link->conf->used_color_bitmap = 0;
+
 	err = ieee80211_set_after_color_change_beacon(link, &changed);
 	if (err) {
 		cfg80211_color_change_aborted_notify(sdata->dev, link->link_id);
@@ -5287,6 +5290,11 @@ ieee80211_color_change(struct wiphy *wiphy, struct net_device *dev,
 		goto out;
 	}
 
+	if (BIT_ULL(params->color) & link_conf->used_color_bitmap) {
+		err = -EINVAL;
+		goto out;
+	}
+
 	err = ieee80211_set_unsol_bcast_probe_resp(sdata,
 						   &params->unsol_bcast_probe_resp,
 						   link, link_conf, &changed);
diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index e84f2b1f..9ddbeb3d 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -211,6 +211,7 @@ enum ieee80211_packet_rx_flags {
  */
 enum ieee80211_rx_flags {
 	IEEE80211_RX_BEACON_REPORTED	= BIT(0),
+	IEEE80211_RX_BEACON_COLOR_COLLISION_CHECKED	= BIT(1),
 };
 
 struct ieee80211_rx_data {
diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index ad40be48..b34460e6 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -3438,13 +3438,12 @@ ieee80211_rx_h_mgmt_check(struct ieee80211_rx_data *rx)
 	    rx->skb->len < IEEE80211_MIN_ACTION_SIZE)
 		return RX_DROP_U_RUNT_ACTION;
 
-	if (rx->sdata->vif.type == NL80211_IFTYPE_AP &&
-	    ieee80211_is_beacon(mgmt->frame_control) &&
-	    !(rx->flags & IEEE80211_RX_BEACON_REPORTED)) {
-		int sig = 0;
+	if (rx->sdata->vif.type != NL80211_IFTYPE_AP ||
+	    !ieee80211_is_beacon(mgmt->frame_control))
+		goto skip_beacon_report;
 
-		/* sw bss color collision detection */
-		ieee80211_rx_check_bss_color_collision(rx);
+	if (!(rx->flags & IEEE80211_RX_BEACON_REPORTED)) {
+		int sig = 0;
 
 		if (ieee80211_hw_check(&rx->local->hw, SIGNAL_DBM) &&
 		    !(status->flag & RX_FLAG_NO_SIGNAL_VAL))
@@ -3457,6 +3456,20 @@ ieee80211_rx_h_mgmt_check(struct ieee80211_rx_data *rx)
 		rx->flags |= IEEE80211_RX_BEACON_REPORTED;
 	}
 
+	if (!(rx->flags & IEEE80211_RX_BEACON_COLOR_COLLISION_CHECKED)) {
+		struct ieee80211_supported_band *sband;
+
+		/* Make sure that the beacon is handled by the correct link */
+		sband = ieee80211_get_link_sband(rx->link);
+		if (!sband || status->band != sband->band)
+			goto skip_beacon_report;
+
+		/* sw bss color collision detection */
+		ieee80211_rx_check_bss_color_collision(rx);
+		rx->flags |= IEEE80211_RX_BEACON_COLOR_COLLISION_CHECKED;
+	}
+
+skip_beacon_report:
 	return ieee80211_drop_unencrypted_mgmt(rx);
 }
 
-- 
2.45.2

